### ISSUE TITLE
Application reports that something went wrong with JobCancellationException

### ISSUE THREAD

**Issue by elizarov** [corona-warn-app/cwa-app-android#1588]
## Describe the bug

See this tweet: https://twitter.com/filippovdmitry/status/1326603774199291904?s=20

## Expected behavior

Canceling an internal operation is normal behavior and should not have been reported as "Something went wrong".

---
Internal Tracking ID:  EXPOSUREAPP-3974

**Comment by d4rken** [comment]
Do you have more details or just the tweet @elizarov ?

I agree that that the cancellation exception is not an error. I'd like to know what was canceled though :thinking:.
There were some earlier bugs where coroutines were launched in the wrong scope (viewmodel scope for UI independent background operations).

btw :heart: Kotlin

**Comment by elizarov** [comment]
Sorry, I don't use the app myself so I cannot really tell how to reproduce it. It should happen in an unlikely course of events when there is an ongoing asynchronous operation and the scope is cancelled for a reason (like a UI action, rotating screen, etc) while it is in process. I've found a few candidate places in the code that might be causing it:

1. https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt#L32-L40
2. https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/InternalExposureNotificationPermissionHelper.kt#L59-L73
3. https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/settings/SettingsTracingFragment.kt#L137-L163
4. https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt#L29-L41
5. https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt#L86-L93

There are more places like this.

What I really wonder is what we can do to help avoid/prevent such bugs from happening in the first place, but I don't have any feasible ideas at the moment.

**Comment by d4rken** [comment]
> https://github.com/corona-warn-app/cwa-app-android/blob/main/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt#L32-L40

This would only throw a cancellation exception if the internal routine fails with an exception though, wouldn't it?
If the scope of the downstream collector is canceled, the collector unsubscribes (cancels?) but the `callbackFlow` will not throw a cancellation exception, at least I couldn't get it to do that:

```kotlin
lifecycleScope.launch {
    coroutineScope() {
        withContext(Dispatchers.Default) {
            defaultTracingStatus.isTracingEnabled.collect { }
            delay(200)
            this.cancel()
        }
    }
}
```

> What I really wonder is what we can do to help avoid/prevent such bugs from happening in the first place, but I don't have any feasible ideas at the moment.

Time, tutorials and documentation? :sweat_smile: I think coroutines have a lower barrier of entry (e.g. in contrast to RxKotlin), but are not free of pitfalls (scoping, cancellation behavior) that might not be directly obvious when starting. So the easy entry barrier means a lot of velocity, and only later when code gets more complex, issues are noticed.

Maybe more opinionated documentation, though a lot of these scoping pitfalls are rather Android specific. Maybe a few LINTing rules could support developers there, but I'm not sure what type of handling/code-patterns the LINT rules should offer, to both help but also not be annoying...

**Comment by dsarkar** [comment]
Dear all: Internal Tracking ID:  EXPOSUREAPP-3974. 

Thanks! Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by heinezen** [comment]
This was fixed in CWA release 1.9 and we have received no further reports on this so far. Therefore, we'll close the issue here.

----
Corona-Warn-App Open Source Team
